src/kfs_frontend/kfs_graph_executor_impl.cpp

```diff
@@ -0,0 +1,1221 @@
+        output->set_name(outputStreamName);
+        output->set_datatype(
+            ovmsPrecisionToKFSPrecision(
+                ovElementTypeToOvmsPrecision(
+                    received.get_element_type())));
+        output->clear_shape();
+        for (const auto& dim : received.get_shape()) {
+            output->add_shape(dim);
+        }
+        response.add_raw_output_contents()->assign(reinterpret_cast<char*>(received.data()), received.get_byte_size());
+        return StatusCode::OK;
+    }
+    HANDLE_PACKET_RECEIVAL_EXCEPTIONS();
+}
+
+template <>
+Status receiveAndSerializePacket<KFSResponse>(const ::mediapipe::Packet& packet, KFSResponse& response, const std::string& outputStreamName) {
+    try {
+        auto received = packet.Get<KFSResponse>();
+        response = std::move(received);
+        return StatusCode::OK;
+    }
+    HANDLE_PACKET_RECEIVAL_EXCEPTIONS();
+}
+
+static KFSDataType convertImageFormatToKFSDataType(const mediapipe::ImageFormat::Format& imageFormat) {
+    static std::unordered_map<mediapipe::ImageFormat::Format, KFSDataType> ImageFormatKFSDatatypeMap{
+        {mediapipe::ImageFormat::GRAY8, "UINT8"},
+        {mediapipe::ImageFormat::SRGB, "UINT8"},
+        {mediapipe::ImageFormat::SRGBA, "UINT8"},
+        {mediapipe::ImageFormat::GRAY16, "UINT16"},
+        {mediapipe::ImageFormat::SRGB48, "UINT16"},
+        {mediapipe::ImageFormat::SRGBA64, "UINT16"},
+        {mediapipe::ImageFormat::VEC32F1, "FP32"},
+        {mediapipe::ImageFormat::VEC32F2, "FP32"}};
+    auto it = ImageFormatKFSDatatypeMap.find(imageFormat);
+    if (it == ImageFormatKFSDatatypeMap.end()) {
+        SPDLOG_DEBUG("Converting Mediapipe::ImageFrame format to KFS datatype failed. Datatype will be set to default - UINT8");
+        return "UINT8";
+    }
+    return it->second;
+}
+
+template <>
+Status receiveAndSerializePacket<mediapipe::ImageFrame>(const ::mediapipe::Packet& packet, KFSResponse& response, const std::string& outputStreamName) {
+    try {
+        const auto& received = packet.Get<mediapipe::ImageFrame>();
+        auto* output = response.add_outputs();
+        output->set_name(outputStreamName);
+        KFSDataType datatype = convertImageFormatToKFSDataType(received.Format());
+        output->set_datatype(datatype);
+        output->clear_shape();
+        output->add_shape(received.Height());
+        output->add_shape(received.Width());
+        output->add_shape(received.NumberOfChannels());
+        cv::Mat image = mediapipe::formats::MatView(&received);
+
+        response.add_raw_output_contents()->assign(reinterpret_cast<char*>(image.data), image.cols * image.rows * image.channels() * image.elemSize1());
+        return StatusCode::OK;
+    }
+    HANDLE_PACKET_RECEIVAL_EXCEPTIONS();
+}
+
+#if (PYTHON_DISABLE == 0)
+template <>
+Status receiveAndSerializePacket<PyObjectWrapper<py::object>>(const ::mediapipe::Packet& packet, KFSResponse& response, const std::string& outputStreamName) {
```

Comment:
Is GIL being held as we enter this function?

Reply:
This code was not modified other than moved from mediapipegraphexecutor.cpp to kfs_graph_executor_impl.cpp

Reply:
It's possible that we had (and still have) a bug then.